### PR TITLE
Detect artists in text to disambiguate matches

### LINKED ISSUE #5: Detect artists while parsing for disambiguation
### Goals
Functionality:
- Find correct artists if present in document
- Prioritize artists by rough proximity to song in document
- Avoid treating homonym songs as duplicates if multiple matching artists are mentioned
  - e.g. if the song "Sorry" is mentioned twice and Justin Bieber and Beyonce are both mentioned

### Strategy
Heuristics and strategies for matching artists:
- Search songs on Spotify while parsing to find small set of candidate artists
- Check enclosing paragraph for candidate artists
- If not present in current paragraph, expand search to rest of document
- Cache artists that are found in paragraph

Takeaways:
- Might make sense to use co-routines per paragraph since there will be network calls
    - However, this would complicate the artist cache since all coroutines would need to read/write access

### Pseudo Code
```
For each song in the text:
  Search song on Spotify
  If many results, filter out hits:
    - that are not tracks
    - by very obscure artists

  For each artist in search results:
    Search for them in current paragraph
    If found:
      - Cache them for later lookup (with location? e.g. in which paragraph?)
      - Add to list of candidate artists

  If no artists found:
    Search rest of document and if found:
      (Same as above)
```

### Testing
Test input file:
> In this case, this song "Sorry" should be detected to be by Justin Bieber because I mentioned him.
>
> Even though, in this separate paragraph, I mention Beyonce, who also has a song named "Sorry". Now, in this case, this song should be detected to be by her. This means that the duplicate detection needs to be smart enough to find this out.
>
> Now I'm in some other paragraph over here, mentioning the artist Kiefer.
>
> The artist that I mentioned in the previous paragraph has a song called "Island". Even though I don't mention him in this paragraph, I don't mention any other artist that has a song by that name, so it'd be nice for the program to detect him (the guy I mentioned in the last paragraph) as a potential artist.

Should give playlist with the following songs:
* Sorry by Justin Bieber
* Sorry by Beyonce
* Island by Kiefer

Currently, the program creates a playlist with two songs:
* Sorry by Justin Bieber
* Island by Florida Georgia Line

### LINKED ISSUE #8: Redefine Song Scrounger API
### Background
While working on #7, I realized that the `document_parser` logic should really be in `song_scrounger` and that `create_playlist` should be directly implemented in `spotify_client`. As its name suggests, `song_scrounger` isn't for playlist management: it's for scrounging songs!

### Summary
The current API conflates the song scrounging and playlist management concepts:
```python3
song_scrounger.create_playlist("input_file.txt", "My New Playlist")
```

Instead, it should look something like:
```python
matches = song_scrounger.find_songs(input_text_file)
spotify_client.create_playlist("My New Playlist", [songs[0].spotify_uri for _, songs in matches.iter()])
```

### PR DESCRIPTION
**Pull request by okjuan**
Resolves #5
Resolves #8

### CODE REVIEW

File: song_scrounger/document_parser.py [hunk 1]
```diff
@@ -1,49 +1,128 @@
...
+        word = word.lower()
+        text = text.lower()
+        return text.find(word) != -1
+
+    def _get_paragraphs(self, text):
+        "Returns non-empty paragraphs with one or more non-whitespace characters."
+        paragraphs = text.split("\n")
+        return [p for p in paragraphs if len(p.strip(" ")) > 0]
+
+    def find_quoted_tokens(self, text):
```
Comment (jack-work) [inline song_scrounger/document_parser.py]: Does `return re.findall("\"([^\"]*)\"", text)` work?  https://docs.python.org/3/library/re.html#re.findall

File: song_scrounger/document_parser.py [hunk 2]
```diff
@@ -1,49 +1,128 @@
...
+                songs_with_mentioned_artists = self.filter_by_mentioned_artist(songs, text)
+                if len(songs_with_mentioned_artists) > 0:
+                    results[song_name] = [song.spotify_uri for song in songs_with_mentioned_artists]
+                else:
+                    results[song_name] = [song.spotify_uri for song in songs]
+        return results
+
+    def filter_by_mentioned_artist(self, songs, text):
+        songs_whose_artists_are_mentioned = set()
+        for song in songs:
```
Comment (jack-work) [inline song_scrounger/document_parser.py]: Cool thing I learned when playing around with list comprehension just now: set comprehension
`return { song if self.is_mentioned(artist, text) else None for song in songs for artist in song.artist }`

No reason to do this with stateful stuff but you can have this anywhere in place of loops which just build a list (slightly more performant too I believe).

File: song_scrounger/document_parser.py [hunk 3]
```diff
@@ -1,49 +1,128 @@
...
+            text (str): e.g. "Hello dear".
+        """
+        word = word.lower()
+        text = text.lower()
+        return text.find(word) != -1
+
+    def _get_paragraphs(self, text):
+        "Returns non-empty paragraphs with one or more non-whitespace characters."
+        paragraphs = text.split("\n")
+        return [p for p in paragraphs if len(p.strip(" ")) > 0]
```
Comment (jack-work) [inline song_scrounger/document_parser.py]: You can also use Regex for this but I can figure it out

File: song_scrounger/document_parser.py [hunk 4]
```diff
@@ -1,49 +1,128 @@
...
\ No newline at end of file
+from .spotify_client import SpotifyClient
+from .models.song import Song
+
+
+class DocumentParser():
+    def __init__(self, spotify_client):
+        self.spotify_client = spotify_client
+
+    async def find_songs(self, text):
```
Comment (IsaacSahle) [inline song_scrounger/document_parser.py]: nit: consider validating text to not be None

File: song_scrounger/document_parser.py (hunk 1 above)
Comment (okjuan) [reply song_scrounger/document_parser.py]: 🙇 🙇‍♂️ 🙇‍♂️

File: song_scrounger/document_parser.py [hunk 5]
```diff
@@ -1,49 +1,128 @@
...
+                of matching songs, empty if no matching artist mentioned.
+        """
+        results = {}
+        paragraphs = self._get_paragraphs(text)
+        for paragraph in paragraphs:
+            song_names = self.find_quoted_tokens(paragraph)
+            for song_name in song_names:
+                songs = await self.search_spotify(song_name)
+                songs_with_mentioned_artists = self.filter_by_mentioned_artist(songs, text)
+                if len(songs_with_mentioned_artists) > 0:
```
Comment (jack-work) [inline song_scrounger/document_parser.py]: If you define this function
`get_songs_in_text_or_default(self, song_name, text):`
`""" gets spotify songs by song_name and returns a set of those which are in the text.  If none are found in the text, it returns all songs.  """`
`songs = await self.search_spotify(song_name)`
`return filter_by_mentioned_artist(self, songs, text) or songs`

You could write this function in one line
`{ song_name: song.spotify_uri for paragraph in self._get_paragraphs(text) for song_name in self.find_quoted_tokens(paragraph) for song in await self.gets_songs_in_text_or_default(song_name, text) }`

File: song_scrounger/document_parser.py [hunk 6]
```diff
@@ -1,49 +1,128 @@
...
+            songs ([Song]).
+        """
+        songs = []
+        tracks = await self.spotify_client.find_track(song)
+        for track in tracks:
+            for artist in track.artists:
+                songs.append(Song(
+                    track.name,
+                    track.uri,
+                    [artist.name for artist in track.artists]
```
Comment (jack-work) [inline song_scrounger/document_parser.py]: Something is not right here...  It looks like you're just adding |track.artists| number of Song objects to the songs list, but you're making it O(|track.artists|^2) by doing that list comprehension in the constructor.

At the very least you could store the list in a variable outside the artist in track.artists loop so you could re-use it (or you could write a function which returns a list of identical Song objects from a track based on the number of artists on the track object), but I'm curious as to why you're adding duplicates here anyway.

Comment (okjuan) [reply song_scrounger/document_parser.py]: Yeah the logic is busted here, I've fixed it locally and pushed up the changes.

File: song_scrounger/document_parser.py (hunk 5 above)
Comment (okjuan) [reply song_scrounger/document_parser.py]: I actually pulled out this if block into a separate function. I'll push up the changes shortly.

I personally don't like using complicated comprehensions, I think they start giving diminishing returns.

File: song_scrounger/document_parser.py (hunk 3 above)
Comment (okjuan) [reply song_scrounger/document_parser.py]: I think this one is probably fine as-is

**Pull request merged**